packages/gateway-client/src/worker/p2p-client/streams.ts

```diff
@@ -0,0 +1,372 @@
+            }
+
+            if (
+                potential.protocol === protocol &&
+                potential.constructor === Constructor
+            ) {
+                this.pool.delete(potential);
+                return potential;
+            }
+        }
+
+        return null;
+    }
+
+    public release() {
+        PooledLobStream.pool.add(this);
+    }
+}
+
+export class RequestStream extends PooledLobStream {
+    public async request(
+        packet: Packet,
+        onChunk?: Callback
+    ): Promise<Packet | null> {
+        if (onChunk) {
+            this.addListener('chunk', onChunk);
+        }
+        this.send(packet);
+        const response = await this.receive();
+        if (onChunk) {
+            this.removeListener('chunk', onChunk);
+        }
+
+        return response;
+    }
+}
+
+export enum WebsocketStreamMessageType {
+    COMMAND = 'COMMAND',
+    MESSAGE = 'MESSAGE',
+    STATUS = 'STATUS',
+}
+
+export enum WebsocketStreamStatus {
+    OPENED = 'OPENED',
+    CLOSED = 'CLOSED',
+    ERROR = 'ERROR',
+}
+
+// TODO: ask joshua how to stop ts complaining about overrides
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+export class WebsocketStream extends LobStream {
+    private statusPort: MessagePort;
+    private commandPort: MessagePort;
+    private messagePort: MessagePort;
+
+    private makePortMessageHandler(type: WebsocketStreamMessageType) {
+        return (event: MessageEvent) => {
+            this.log.debug(`${type}`, event.data);
+            const message = this.encodeMessageToPacket(
+                type,
+                Buffer.from(event.data)
+            );
+            this.send(message);
+        };
+    }
+
+    constructor(libp2pStream: Stream, ports: MessagePort[]) {
+        super(libp2pStream, ports);
+        this.statusPort = ports[0];
+        this.messagePort = ports[1];
+        this.commandPort = ports[2];
+        this.startSending();
+        this.startReceiving();
+    }
+
+    private startSending() {
+        this.commandPort.onmessage = this.makePortMessageHandler(
+            WebsocketStreamMessageType.COMMAND
+        );
+        this.messagePort.onmessage = this.makePortMessageHandler(
+            WebsocketStreamMessageType.MESSAGE
+        );
+        this.statusPort.onmessage = this.makePortMessageHandler(
+            WebsocketStreamMessageType.STATUS
+        );
+    }
+
+    private async startReceiving() {
+        let packet = null;
+
+        while (this.isOpen && (packet = await this.receive()) !== null) {
+            this.dispatch(packet);
+        }
+
+        this.dispatchStatus(WebsocketStreamStatus.CLOSED);
+    }
+
+    private encodeMessageToPacket(
+        type: WebsocketStreamMessageType,
+        buffer: Buffer
+    ): Packet {
+        return encode({ type, bodyLength: buffer.byteLength }, buffer);
+    }
+
+    private getClientPort(type: WebsocketStreamMessageType): MessagePort {
+        switch (type) {
```

Comment:
It would be better to use an exhaustive type here so that any missing types throw compile errors instead of runtime errors.

You can do this by creating a new `Record<WebsocketSreamMessageType, MessagePort>` instance property that gets assigned in the constructor. You'll still need to keep the invalid type error since you're taking in input from the server, but doing this will avoid the mistake of leaving out new enum values.